snap/naming/tag.go

```diff
@@ -0,0 +1,125 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package naming
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+var errInvalidSecurityTag = errors.New("invalid security tag")
+
+// ParsedSecurityTag describes a parsed snap security tag.
+type ParsedSecurityTag interface {
```

Comment:
I haven't seen examples of use, but `ParseSecurityTag()` returning an error and combining that with type assertion I can see how it gets cumbersome.

Would it be more convenient to have:
```
type SecurityTag interface {
	// String returns the entire security tag.
	String() string

	// InstanceName returns the snap name and instance key.
	InstanceName() string
	// AppName returns the name of the application if the tag belongs
	// to an application, otherwise an empty string is returned.
	AppName() string
	// HookName returns the name of the hook if the tag belongs to one,
	// otherwise an empty string is returned.
	HookName() string
}
```

OTOH if you compare:
```
tag, err := ParseSecurityTag("snap.foo.bar")
if err != nil {
	return err
}
hookTag, ok := tag.(HookSecurityTag)
if !ok {
	return fmt.Errorf("not a hook tag")
}
```
and

```
tag, err := ParseSecurityTag("snap.foo.bar")
if err != nil {
	return err
}
hook, := tag.HookName()
if hook != "" {
	return fmt.Errorf("not a hook tag")
}
```
The complexity is similar?

Reply:
Yeah, I considered this and decided to go with the direct dedicated interface. There are other security tags that are not supported yet and having them distinct may help to explain they belong to different use cases.

Having said that, my only usage of this code (in refresh-app-awareness-v2) does not use type assertions as ll I need is the snap instance name.

I don't have very strong opinions on this so if it comes out that bundling into one interface is more practical we can move there.

Reply:
I think I prefer typed approach from the PR, it forces correctness and makes it harder to forget a == "" check.
Perhaps two simple wrappers around the first snippet could simplify some use cases where a specific type is expected, e.g:
```
func ParseHookSecurityTag(tag string) (error, HookSecurityTag) {
...
}

func ParseAppSecurityTag(tag string) (error, AppSecurityTag) {
...
}
```

Reply:
I like this idea. I'll use it when this code spreads around and there are more use cases that need to parse the tags.

Reply:
+1 to @stolowski's suggestion